assets/js/src/settings/store/actions.ts

```diff
@@ -25,3 +30,138 @@ export function* saveSettings() {
   yield { type: 'TRACK_SETTINGS_SAVED' };
   return { type: 'SAVE_DONE' };
 }
+
+export function* verifyMssKey(key: string, activateMssIfKeyValid: boolean) {
+  const { success, error, res } = yield {
+    type: 'CALL_API',
+    endpoint: 'services',
+    action: 'checkMSSKey',
+    data: { key },
+  };
+  if (!success) {
+    return updateKeyActivationState({
+      mssStatus: 'invalid',
+      mssMessage: error.join(' ') || null,
+    });
+  }
+  const fields: Partial<KeyActivationState> = {
+    mssMessage: res.data.message || null,
+  };
+  if (activateMssIfKeyValid) {
+    const call = yield {
+      type: 'CALL_API',
+      endpoint: 'settings',
+      action: 'set',
+      data: {
+        mta_group: 'mailpoet',
+        mta: { method: 'MailPoet', mailpoet_api_key: key },
+        signup_confirmation: { enabled: '1' },
+      },
+    };
+    if (!call.success) {
+      fields.mssStatus = 'valid_mss_not_active';
+    } else {
+      yield setSetting(['mta_group'], 'mailpoet');
+      yield setSetting(['mta', 'method'], 'MailPoet');
+      yield setSetting(['mta', 'mailpoet_api_key'], key);
+      yield setSetting(['signup_confirmation', 'enabled'], '1');
+      fields.mssStatus = 'valid_mss_active';
+    }
+  } else {
+    fields.mssStatus = 'valid_mss_not_active';
+  }
+  yield updateKeyActivationState(fields);
+  return fields.mssStatus;
+}
+
+export function* verifyPremiumKey(key: string) {
+  const { res, success, error } = yield {
+    type: 'CALL_API',
+    endpoint: 'services',
+    action: 'checkPremiumKey',
+    data: { key },
+  };
+  if (!success) {
+    MailPoet.trackEvent(
+      'User has failed to validate a Premium key',
+      {
+        'MailPoet Free version': MailPoet.version,
+        'Premium plugin is active': !!MailPoet.premiumVersion,
+      }
+    );
+    return updateKeyActivationState({
+      premiumStatus: 'invalid',
+      premiumMessage: error.join(' ') || null,
+    });
+  }
+
+  yield updateKeyActivationState({ premiumMessage: null });
+  // install/activate Premium plugin
+  let pluginInstalled = res.meta.premium_plugin_installed;
+  let pluginActive = res.meta.premium_plugin_active;
+
+  if (!pluginInstalled) {
+    const actions = installPremiumPlugin();
+    let action = actions.next();
+    while (!action.done) {
+      yield action.value;
+      action = actions.next();
+    }
+    pluginInstalled = action.value;
```

Comment:
Why does this need to be iterated? Could we just yield it as with "activatePremiumPlugin" below?

Reply:
You are right, my bad.